### PR TITLE
Check graph invariants during tests

### PR DESCRIPTION
**Pull request by wchargin**
Summary:
Each of the invariants listed at the top of the `Graph` class is now
explicitly checked by `checkInvariants`, which is called at the end of
each `Graph` method during tests only. This is powerful: it means that
not only do our tests for `Graph` test the graph, but also any tests
that depend on `Graph`—e.g., plugin code—will give us extra invariant
testing on `Graph`. As noted in a comment, if this becomes bad for
performance, we can blacklist expensive tests or whitelist tests that we
care about.

A graph method may assume that the graph invariants hold before the
method is invoked. Within the body of a graph method, invariants may be
violated, but the method must ensure that the invariants hold
immediately before it returns or yields. A consequence of this is that
if a graph function internally calls a public function (e.g., `addEdge`
might call `hasNode` to check that the source and destination exist),
then it must ensure that the invariants hold before the internal call.
This is not an “implementation detail” or “caveat”; it is simply part of
the interface of public functions. It is legal and reasonable for
private helper functions to explicitly not expect or not guarantee that
particular invariants hold, and in this case the exception should be
documented. (This is not yet the case in any of our code.)

Finally, note that the `checkInvariants` method should not call any
public methods, because those methods in turn call `checkInvariants`. If
this becomes a huge pain, we can look into implementing some kind of
“only check invariants if the invariants are not actively being
checked”, but I’d much rather not do so if we don’t have to.

Test Plan:
Running `yarn coverage` indicates that each of the failure cases is
verified. In principle, I’d be willing to add a test that parses the
source code for `graph.js` and verifies that each `return`, `yield`, or
implicit return is preceded by an invariant check. But I don’t really
want to implement that right now.

wchargin-branch: automatic-invariants

### CODE REVIEW

**Comment by wchargin** [comment]
> It appears this code verifies the following properties: […] Should
> this be documented in the list of invariants being checked by this
> code?

Thanks for pointing this out. You’re right, and this is indicative of a
more general problem. These invariants are easy to read, but hard to
understand. It’s difficult to see why the pieces fit together to
encapsulate all (and only) the conditions that we want, or, conversely,
to notice when they don’t.

I’ve rewritten the invariants to make them clearer. I’ve reduce the use
of TFAEs, which I think have introduced some subtleties. Now, we have
one invariant for each property of the class, describing what it is that
an element of that property implies about other elements and properties.
This also lends itself to a nice implementation.

Here are the new proposed invariants:

```js
  // Definition. A node `n` is in the graph if `_nodes.has(n)`.
  //
  // Definition. An edge `e` is in the graph if `_edges.has(e.address)`.
  //
  // Definition. A *logical value* is an equivalence class of ECMAScript
  // values modulo deep equality (or, from context, an element of such a
  // class).
  //
  // Invariant 1. For a node `n`, if `n` is in the graph, then
  // `_inEdges.has(n)` and `_outEdges.has(n)`.
  //
  // Invariant 2. For an edge `e`, if `e` is in the graph, then:
  //  1. `e.src` is in the graph;
  //  2. `e.dst` is in the graph;
  //  3. `edges.get(e.address)` is deep-equal to `e`;
  //  4. `_inEdges.get(e.dst)` contains `e`; and
  //  5. `_outEdges.get(e.src)` contains `e`.
  //
  // Invariant 3. Suppose that `_inEdges.has(n)` and, let `es` be
  // `_inEdges.get(n)`. Then
  //  1. `n` is in the graph;
  //  2. `es` contains any logical value at most once;
  //  3. if `es` contains `e`, then `e` is in the graph; and
  //  4. if `es` contains `e`, then `e.dst === n`.
  //
  // Invariant 4. Suppose that `_outEdges.has(n)` and, let `es` be
  // `_outEdges.get(n)`. Then
  //  1. `n` is in the graph;
  //  2. `es` contains any logical value at most once;
  //  3. if `es` contains `e`, then `e` is in the graph; and
  //  4. if `es` contains `e`, then `e.src === n`.
```

What do you think?

(Note: There may be errors in the above; I’ve checked them only briefly.
I’ll review your outstanding PRs now. :-) )

**Comment by decentralion** [comment]
Looks good overall. Two comments:

1. I'd like to have the initial invariants about `inEdges` and `outEdges` also declare the type signature (contains an array of zero-or-more edges). Just so the invariants are a little more self contained. (It might be worth giving a bit more info about what an edge is in the definition.)

2. I think we can improve the performance of this invariant checking: using a counting argument, if we verify for every valid node that it is present in `_inEdges`, and every valid edge's dst appears in `_inEdges` then:
* we can count the total number of keys in `_inEdges`, to verify that there are no extra nodes appearing
* we can count the total number of entries in `_inEdges`, to verify that there are no extra edges appearing
And the same for `_outEdges`. computationally, this saves us all of the work in section 3 and 4.
As discussed offline, this should include a proof to be persuasive, and it's not yet a perf bottleneck, so we can punt on that optimization and just implement the invariants you've laid out here.

File: src/v3/core/graph.js [hunk 1]
```diff
@@ -72,6 +72,118 @@ export class Graph {
   _inEdges: Map<NodeAddressT, Edge[]>;
   _outEdges: Map<NodeAddressT, Edge[]>;
 
+  checkInvariants() {
+    // Invariant: For a node `n`, the following are equivalent:
```
[the remark this answers is not in the document]
Comment (wchargin) [reply src/v3/core/graph.js]: Done.

File: src/v3/core/graph.js [hunk 2]
```diff
@@ -72,6 +72,118 @@ export class Graph {
...
+      if (!this._nodes.has(edge.dst)) {
+        throw new Error(`missing dst for edge: ${edgeToString(edge)}`);
+      }
+    }
+
+    // Invariant: For an edge `e`, if `e.dst` and `e.src` are in the
+    // graph, the following are equivalent:
+    //  1. `e` is in the graph;
+    //  2. `_inEdges.get(e.dst)` contains `e`;
+    //  3. `_inEdges.get(e.dst)` contains `e` exactly once;
```
[the remark this answers is not in the document]
Comment (wchargin) [reply src/v3/core/graph.js]: :-)

**Pull request merged**